Allow dynamic proxy_ssl_name. Fixes #3172

mreiche opened the pull request:
Fixes https://forum.opnsense.org/index.php?topic=17637.0
See more details in linked issue https://github.com/opnsense/plugins/issues/3172

kulikov-a commented:
Hi!
IMHO it looks reasonable and allows to reduce the required number of created servers just for the sake of passing the correct SNI hostname.
Maybe then it is worth updating the help text at https://github.com/opnsense/plugins/blob/548566b52db2e7a5799bb67530f88086b63546b6/www/nginx/src/opnsense/mvc/app/controllers/OPNsense/Nginx/forms/upstream.xml#L43 too?

fabianfrz commented:
This behaviour would be correct for `http->nginx->https` (http://nginx.org/en/docs/http/ngx_http_core_module.html#var_host)

however in case of `https->nginx->https`, it would be more correct to forward the server name we get via TLS: http://nginx.org/en/docs/http/ngx_http_ssl_module.html#var_ssl_server_name as otherwise it would allow host spoofing.

The best way to implement that is by using a map together with the protocol (https vs. http).

Also keep in mind that streams (TCP and UDP) are there as well.

mreiche commented:
> Hi! IMHO it looks reasonable and allows to reduce the required number of created servers just for the sake of passing the correct SNI hostname. Maybe then it is worth updating the help text at

I've updated the documentation to make more clear, what to expect to be changed. The documentation was a bit confusing for me before. I also added names of the variables to be set. I don't now if there should be another markdown to be used since I did not find another example anywhere in the help texts of other options.

fichtner commented on www/nginx/src/opnsense/mvc/app/controllers/OPNsense/Nginx/forms/upstream.xml:
```diff
@@ -40,7 +40,7 @@
     <id>upstream.tls_name_override</id>
     <label>TLS: Servername override</label>
-    <help>Use another hostname for SNI and certificate validation.</help>
+    <help>Use another hostname for SNI and certificate validation as $proxy_ssl_name.</help>
```
Do we have to explain this with variables?

mreiche replied:
For me personally, it helps a lot to figure out, what the setting does. Otherwise we have to "guess" what changed in `nginx.conf`.

kulikov-a replied:
may be something like "Override the server name used to verify the Upstream certificate and to be passed to Upstream through SNI (proxy_ssl_name). By default the SNI name from request or $host variable is used"

mreiche replied:
@fichtner Please provide me your preference.

fabianfrz replied:
How about "Force a specific hostname for the backend connection instead of passing the hostname from the downstream connection to the upstream connection."

fichtner commented on www/nginx/src/opnsense/mvc/app/controllers/OPNsense/Nginx/forms/location.xml:
```diff
@@ -255,7 +255,7 @@
     <type>checkbox</type>
     <advanced>true</advanced>
-    <help>Check this box, if you want the client SNI header to be used instead of your backend hostname. This settings overrides the configured hostname in the upstream configuration.</help>
+    <help>Check this box, if you want the client SNI header ($ssl_server_name) to be used as $proxy_ssl_name. Can be overriden in the upstream's hostname configuration.</help>
```
This declines the quality of the help text actually. I don't have a suggestion for lack of knowledge but I don't want this in as is.

mreiche replied:
Where else should I document these important technical details?

kulikov-a replied:
"Enables passing of the server name through TLS Server Name Indication (SNI) when establishing a connection with Upstream (proxy_ssl_server_name)"?

mreiche replied:
@fichtner Please provide me your preference.

fabianfrz replied:
How about this:

... if you want to passthrough the hostname you get from the downstream to the upstream connection. ....

... Please note, that TLS server certificates are validated against this servername. If the server name passed by the client leads to different processing on the upstream server (i. e. different virtual host is choosen), this may cause security leaks.

mreiche commented:
> The best way to implement that is by using a map together with the protocol (https vs. http).

I've implemented the switch based on the backend's TLS option. I hope that is enough. Otherwise I need help to find the proper option.


> Also keep in mind that streams (TCP and UDP) are there as well.

In our case, we have multiple cluster backends routed by the same reverse proxy. Thats why we need HTTPs forwarding and no TCP-forwarding and a dynamic `$proxy_ssl_name` there.

kulikov-a commented:
@mreiche 
imho checking the `upstream.tls_enable` value not what @fabianfrz meant. 
Let's assume that although we communicate with Upstream using TLS and SNI, the Server allows connection with a plain HTTP or TLS without SNI ("old clients")? in this case `$ssl_server_name` will be empty.
In my opinion, it is more correct in this case to add such a mapping:
`
map $ssl_server_name $upstream_sni_name {
    default $ssl_server_name;
    ''      $host;
}
`
and use `$upstream_sni_name` for proxy_ssl_name default.

@fabianfrz , if I understand correctly, do you mind if we hardcode this in a http block?

short note on the use of a  `$ssl_server_name` variable: imho with all that "http/2 connection reuse" stuff, sni hostname and request Host header mismatch is not so uncommon. and although I believe that "sni hostname forwarding" is the most correct way, for some it may come as a surprise (first of all, will the backend and its certificates be ready for this?).
may be we need some choice here? (first: $upstream_sni_name (if present) or $host, second: $host only?)

kulikov-a commented on www/nginx/src/opnsense/service/templates/OPNsense/Nginx/nginx.conf:
```diff
@@ -33,4 +33,8 @@ stream {
 # }
+
+# Map used in location.conf for proxy_ssl_name
+map $ssl_server_name $upstream_sni_name { default $ssl_server_name; '' $host; }
```
http.conf template already have map directive with a proper nginx.conf style (https://github.com/opnsense/plugins/blob/548566b52db2e7a5799bb67530f88086b63546b6/www/nginx/src/opnsense/service/templates/OPNsense/Nginx/http.conf#L50). i would place it there, after the $connection_upgrade mapping

kulikov-a reviewed (approved):
lgfm

fichtner commented:
Merged, thanks!
